### PR TITLE
[WIP] Automatically detect interpreters (alternate approach)

### PR DESCRIPTION
**Pull request by agaffney**
##### SUMMARY

This commit adds an action plugin called 'gather_interpreters' and runs it automatically as part of play fact gathering. This plugin uses 'which' (on platforms where it's supported) to find the full path of a list of possible interpreter names and sets the appropriate 'ansible_x_interpreter' var.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
interpreter path

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (gather_interpreters 1a60f99df2) last updated 2018/07/14 10:49:22 (GMT -500)
```

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

**Comment by sivel** [comment]
Overall this is the approach that I was discussing at some point.  The biggest difference is that I think instead of trying to use `which` I was recommending just trying to exec the python versions with `-V` or something.  Just figured I'd mention it.

And instead of an action I was using a new block, calling the raw module in a block/rescue/finally manner.

Anyway, just ideas.

**Comment by agaffney** [comment]
This implementation can be used for interpreters other than python by adding them to the list, and trying to parse version numbers from a bunch of arbitrary interpreters doesn't sound pleasant.

I opted for an action plugin so that the action could be performed again arbitrarily, like is currently possible for `setup`.

**Comment by sivel** [comment]
> trying to parse version numbers from a bunch of arbitrary interpreters doesn't sound pleasant

Oh, I had no intention of actually parsing the response.  I was just using `-V` as it would cause the interpreter to not open, just print the version and exit.

File: lib/ansible/plugins/shell/sh.py [hunk 1]
```diff
@@ -40,6 +40,12 @@ class ShellModule(ShellBase):
     _SHELL_GROUP_LEFT = '('
     _SHELL_GROUP_RIGHT = ')'
 
+    def find_binary(self, bins):
+        if not type(bins) == list:
```
Comment (sivel) [inline lib/ansible/plugins/shell/sh.py]: This should be an isinstance check

Comment (agaffney) [reply lib/ansible/plugins/shell/sh.py]: Fixed

**Comment by agaffney** [comment]
I've realized this this implementation implicitly relies on facts being injected as vars, which means that it may be necessary to inject the vars into VarsManager directly.

**Comment by mkrizek** [comment]
Superseded by https://github.com/ansible/ansible/pull/50163

**Pull request closed without merging**